cpp/src/arrow/compute/exec/source_node.cc

```diff
@@ -196,17 +186,16 @@ struct SourceNode : ExecNode {
   void StopProducing() override {
     std::unique_lock<std::mutex> lock(mutex_);
     stop_requested_ = true;
+    if (!started_) finished_.MarkFinished();
```

Comment:
If `started_` is false wouldn't that have to mean that someone called `StopProducing` before they called `StartProducing`?  What kind of situation are you trying to guard against here?

Reply:
Yes, that's what I'm guarding against. Part 2 of the refactor will fix this weirdness, but essentially if we encounter an error during `StartProducing`, we call `StopProducing` of every node that has been started. These nodes then call `StopProducing` on their inputs, which ends up propagating down to the `SourceNode`. 

Part2 will get rid of this `StopProducing` spam by having each node have an `Abort` method which will be called once by the plan. Nodes will not be calling `Abort` on each other.